keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,498 @@
+Solutions like Intel's Balloons Policy can be deployed to separate infrastructure and workload into different CPU pools but they require extra software, additional tuning and reduced CPU pool size could affect performance of multi-threaded processes.
+
+### Goals
+ * Align scheduler and node view for Node Allocatable (total - reserved).
+ * Ensure `reservedSystemCPUs` is only used by system daemons or interrupt processing not by workloads.
+ * Ensure no breaking changes for the `static` policy of `CPUManager`.
+
+### Non-Goals
+
+## Proposal
+
+We propose to add a new `CPUManager` policy option called `strict-cpu-reservation` to the `static` policy of `CPUManager`.
+When this policy option is enabled, we remove the reserved cores from the list of all available cores at the stage of calculation DefaultCPUSet. As a result, burstable and best-effort containers are launched with a cpuset in which the reserved cores are excluded.
+
+### User Stories (Optional)
+
+#### Story 1
+To protect latency of workload, systemd daemons including irqbalance daemon are commonly constrained to the reserved CPUs.
+Burstable and best-effort pods (and guaranteed pods with fractional CPU requests) running on the reserved CPUs causes CPU throttling for infrastructure services which results in poor system response time which in turn hits back on workload response time.
+This issue is particularly bad in all-in-one deployments where workloads are placed on combined master+worker+storage nodes.
+
+#### Story 2
+Silently allowing workloads running on the reserved CPUs makes benchmarking infrastructure and workloads both inaccurate.
+
+### Risks and Mitigations
+
+The feature is isolated to a specific policy option `strict-cpu-reservation` under `cpuManagerPolicyOptions` and is protected by feature gate `CPUManagerPolicyAlphaOptions` or `CPUManagerPolicyBetaOptions` before the feature graduates to `Stable` i.e. enabled by default.
+
+Concern for feature impact on best-effort workloads, the workloads that do not have resource requests, is brought up.
+
+Kube-scheduler schedules pods on node allocatable (total - reserved). For best-effort pods, kube-scheduler uses default request values when scoring the nodes, see https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/util/pod_resources.go#L32 and https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/noderesources/resource_allocation.go#L123, but the scheduler does not use the default request values when fitting the nodes i.e. best-effort pods are always admitted.
+
+The concern is, when the feature graduates to `Stable`, it will be enabled by default, best-effort workloads could be starved on the node when the node runs out of non-reserved CPU cores.
```

Comment:
Any reason the feature can't be made opt in even when it graduates to Stable?

Reply:
This is probably my fault, my comments could have been lead to this wording, apologies about that.
The following all need to be true at the same time to trigger the possible Best Effort pod starvation (Burstable and Guaranteed pods won't be affected)
- the feature gate must be enabled
- the cpumanager option must be set
- the node must be saturated with guaranteed pods all of them requiring exclusive cpus.

it's a nice case indeed and I think we reached agreement to add metrics for alpha stage so we can evaluate the real impacts of this scenario and postpone any further action to beta or following